### PR TITLE
Fix warnings and error building swift/SILOptimizer on Windows with MSVC

### PR DESCRIPTION
**Pull request by hughbe**
- Control path uncovered warning
- Non accessible method error
- Can't friend class error
- Sign conversion warnings
- Internal stdlib assertitions when compiling swift code

### CODE REVIEW

File: lib/SILOptimizer/Analysis/EscapeAnalysis.cpp [hunk 1]
```diff
@@ -789,6 +789,9 @@ namespace llvm {
         case CGForDotView::PointsTo: return "";
         case CGForDotView::Deferred: return "color=\"gray\"";
       }
+
+
+        llvm_unreachable("Unhandled CGForDotView in switch.");
```
Comment (rudkx) [inline lib/SILOptimizer/Analysis/EscapeAnalysis.cpp]: The indentation is off here.

**Review by rudkx — COMMENTED** [review]
Otherwise LGTM but @eeckstein should probably take a quick look.

File: include/swift/SILOptimizer/Analysis/IVAnalysis.h [hunk 2]
```diff
@@ -26,7 +26,6 @@ namespace swift {
 class IVInfo : public SCCVisitor<IVInfo> {
 public:
   typedef llvm::SmallVectorImpl<ValueBase *> SCCType;
-  friend class SCCVisitor;
```
Comment (jrose-apple) [inline include/swift/SILOptimizer/Analysis/IVAnalysis.h]: I think the right way to fix this is just `friend SCCVisitor;`, not to make the `visit` method public.

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 3]
```diff
@@ -1107,9 +1107,15 @@ processFunction(SILFunction &F, bool EnableDiagnostics,
...
-                                                FoldedUsers.size());
-    if (!UserArray.empty()) {
+    // MSVC hits a debug assertion inside the C++ stdlib ("vector iterator
+    // not dereferencable") if FoldedUsers is empty.
+    ArrayRef<SILInstruction *> UserArray;
+    if (FoldedUsers.size() == 0) {
+      UserArray = ArrayRef<SILInstruction *>();
+    }
+    else {
+      UserArray = ArrayRef<SILInstruction *>(&*FoldedUsers.begin(),
```
Comment (jrose-apple) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: SetVector has a `getArrayRef` method you can use instead.

File: lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp [hunk 4]
```diff
@@ -758,7 +758,7 @@ SILCombiner::visitInjectEnumAddrInst(InjectEnumAddrInst *IEAI) {
       SILBasicBlock *DefaultBB = nullptr;
 
       if (SEI->hasDefault()) {
-        auto *IL = B.createIntegerLiteral(SEI->getLoc(), IntTy, APInt(32, SEI->getNumCases(), false));
+        auto *IL = B.createIntegerLiteral(SEI->getLoc(), IntTy, APInt(32, static_cast<uint64_t>(SEI->getNumCases()), false));
```
Comment (jrose-apple) [inline lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp]: Please wrap to 80 columns.

**Comment by eeckstein** [comment]
LGTM

File: include/swift/SILOptimizer/Analysis/IVAnalysis.h (hunk 2 above)
Comment (hughbe) [inline include/swift/SILOptimizer/Analysis/IVAnalysis.h]: I get an error: “class template has already been declared as a non-class template”

I think the fix is to change this to `friend calss SCCVisitor<IVInfo>` which compiles in VS. I'm running a full build now to make sure

File: lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp (hunk 4 above)
Comment (hughbe) [inline lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp]: fixed

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp (hunk 3 above)
Comment (hughbe) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: good idea, I also removed the MSVC comment

File: lib/SILOptimizer/Analysis/EscapeAnalysis.cpp (hunk 1 above)
Comment (hughbe) [inline lib/SILOptimizer/Analysis/EscapeAnalysis.cpp]: Thanks, I also removed the double new line

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test and merge

**Comment by hughbe** [comment]
This didn't get merged either although tests passed!

**Pull request merged**